packages/runtime-core/src/componentProps.ts

```diff
@@ -317,16 +318,21 @@ function resolvePropValue(
   value: unknown,
   instance: ComponentInternalInstance
 ) {
+  const { propsDefaultValue } = instance
   const opt = options[key]
   if (opt != null) {
     const hasDefault = hasOwn(opt, 'default')
     // default values
     if (hasDefault && value === undefined) {
       const defaultValue = opt.default
       if (opt.type !== Function && isFunction(defaultValue)) {
-        setCurrentInstance(instance)
-        value = defaultValue(props)
-        setCurrentInstance(null)
+        if (propsDefaultValue[key] !== undefined) {
```

Comment:
We could also do
```suggestion
        if (key in propsDefaultValue) {
```

It's shorter which probably doesn't matter here but it also avoids the `else` statement.

It's also 10% faster in Edge, Chrome and Firefox **but 90% slower on Safari** 😆 . So not sure if my suggestion is worth applying (https://jsbench.me/u7kmn8d67h/1)